index.bs

```diff
@@ -1288,100 +1344,133 @@ partial interface ModelBuilder {
-  "default",
-  // Prioritizes execution speed over other considerations e.g. power consumption
-  "high-performance",
-  // Prioritizes power consumption over other considerations e.g. execution speed
-  "low-power"
+[SecureContext, Exposed=Window]
+interface MLModel {
+  Promise<MLCompilation> compile();
 };
+</script>
+
+## Compilation ## {#api-mlcompilation}
+The {{MLCompilation}} interface represents a compiled version of the computational graph of a {{MLModel}} according to the compilation options. A compilation of a graph once constructed is immutable and cannot be subsequently changed.
 
-dictionary CompilationOptions {
-  // Compilation preference as related to power consumption level
-  PowerPreference powerPreference = "default";
+<script type=idl>
+// WebGL specification:
+// (https://www.khronos.org/registry/webgl/specs/latest/1.0/#5.9)
+[SecureContext, Exposed=Window]
+interface WebGLTexture {
+  // ...
 };
 
+// WebGPU specification:
+// (https://gpuweb.github.io/gpuweb/#texture-interface)
 [SecureContext, Exposed=Window]
-interface Model {
-  Promise<Compilation> compile(optional CompilationOptions options = {});
+interface GPUTexture {
+  // ...
 };
-</script>
 
-## Compilation ## {#api-compilation}
-The {{Compilation}} interface represents a compiled version of the computational graph of a {{Model}} according to the compilation options. A compilation of a graph once constructed is immutable and cannot be subsequently changed.
+dictionary MLInput {
+  required ArrayBufferView bufferView;
+  sequence<long> dimensions;
+};
 
-<script type=idl>
-dictionary Input {
-  required ArrayBufferView buffer;
+dictionary MLOutput {
+  ArrayBufferView bufferView;
   sequence<long> dimensions;
 };
 
-dictionary Output {
-  ArrayBufferView buffer;
+dictionary MLWebGLInput {
+  required WebGLTexture glTexture;
   sequence<long> dimensions;
 };
 
-typedef record<DOMString, Input> NamedInputs;
-typedef record<DOMString, Output> NamedOutputs;
+dictionary MLWebGLOutput {
+  WebGLTexture glTexture;
+  sequence<long> dimensions;
+};
+
+dictionary MLGPUInput {
+  required GPUTexture gpuTexture;
+  sequence<long> dimensions;
+};
+
+dictionary MLGPUOutput {
+  GPUTexture gpuTexture;
+  sequence<long> dimensions;
+};
+
+typedef record<DOMString, MLInput> MLNamedInputs;
+typedef record<DOMString, MLOutput> MLNamedOutputs;
+
+typedef record<DOMString, MLWebGLInput> MLNamedWebGLInputs;
+typedef record<DOMString, MLWebGLOutput> MLNamedWebGLOutputs;
+
+typedef record<DOMString, MLGPUInput> MLNamedGPUInputs;
+typedef record<DOMString, MLGPUOutput> MLNamedGPUOutputs;
 
 [SecureContext, Exposed=Window]
-interface Compilation {
-  Promise<NamedOutputs> compute(NamedInputs inputs, optional NamedOutputs outputs = {});
+interface MLCompilation {
+  Promise<MLNamedOutputs> compute(MLNamedInputs inputs, 
+                                  optional MLNamedOutputs outputs = {});
+
+  Promise<MLNamedWebGLOutputs> compute(MLNamedWebGLInputs glInputs, 
```

Comment:
Should the `MLCompilation` compiled for `WebGLRenderingContext` only accept `WebGLTexture` as input and output? Otherwise, there might be expensive data transferring between devices?

Reply:
Yes, it will fail. Cross-device, cross-adapter support should not occur at this level. If the context is created with one device/adapter and the input is a resource from another, the API should fail. The callers of WebNN needs to ensure that the all the device-dependent resources are from the device that is bound to the context.

Reply:
It makes sense. We may need to specify this behavior in the following PR.